### PR TITLE
Support ordered memory pool reclaim

### PR DESCRIPTION
**Pull request by xiaoxmeng**
In Meta internal arbitration and spilling shadowing test, we found that without
enforcing reclaim order from a plan node, then the arbitrator always reclaim
memory follow the memory pool hierarchy, that is, always first reclaim from the
first child, second and so on. This might generates small files as the first operator
has less amount of in-memory state than the other which is not efficient as for
aggregation, we want to reduce the number of files as each file is a sorted run.

This PR optimizes this by sort the memory pool based on its reserved memory
before the actual memory reclaimation.

### CODE REVIEW

**Comment by mbasmanova** [comment]
@xiaoxmeng Ci is red.

File: velox/common/memory/MemoryArbitrator.cpp [hunk 1]
```diff
@@ -187,20 +187,41 @@ uint64_t MemoryReclaimer::reclaim(MemoryPool* pool, uint64_t targetBytes) {
   if (pool->kind() == MemoryPool::Kind::kLeaf) {
     return 0;
   }
-  // TODO: add to sort the child memory pools based on the reclaimable bytes
-  // before memory reclamation.
+
+  // Sort the child pools based on their reserved memory and reclaim from the
```
Comment (mbasmanova) [inline velox/common/memory/MemoryArbitrator.cpp:191]: It feels that we may want to iterate on this logic and maybe customize it based on application. I wonder if it makes sense to extract this logic into a separate class. Also, the TODO said to sort by "reclaimable bytes", not "reserved bytes". Reclaiming from the operator with most bytes available to reclaim makes more sense, no?

File: velox/exec/Driver.h [hunk 2]
```diff
@@ -403,13 +403,15 @@ class Driver : public std::enable_shared_from_this<Driver> {
   friend struct DriverFactory;
 };
 
+#ifdef VELOX_ENABLE_BACKWARD_COMPATIBILITY
```
Comment (mbasmanova) [inline velox/exec/Driver.h:406]: Unrelated change?

File: velox/exec/MemoryReclaimer.cpp [hunk 3]
```diff
@@ -20,11 +20,11 @@
 #include "velox/exec/Task.h"
 
 namespace facebook::velox::exec {
-std::unique_ptr<memory::MemoryReclaimer> DefaultMemoryReclaimer::create() {
-  return std::unique_ptr<MemoryReclaimer>(new DefaultMemoryReclaimer());
+std::unique_ptr<memory::MemoryReclaimer> MemoryReclaimer::create() {
```
Comment (mbasmanova) [inline velox/exec/MemoryReclaimer.cpp:23]: Unrelated changes?

File: velox/common/memory/MemoryArbitrator.cpp (hunk 1 above)
Comment (xiaoxmeng) [reply velox/common/memory/MemoryArbitrator.cpp:191]: The reserved bytes is more relevant here because we can also release unused reservation at leaf level. My previous version adds a ordered reclaim flag in memory reclaimer to customize the logic.

File: velox/exec/Driver.h (hunk 2 above)
Comment (xiaoxmeng) [reply velox/exec/Driver.h:406]: I have separated these changes in a separate pr.

**Comment by xiaoxmeng** [comment]
> @xiaoxmeng Ci is red.

Will fix. Thanks!

**Review by mbasmanova — APPROVED** [review]
Thanks.
